src/main/java/com/conveyal/datatools/manager/models/transform/NormalizeFieldTransformation.java

```diff
@@ -0,0 +1,354 @@
+            result = substitution.replace(result);
+        }
+
+        return result;
+    }
+
+    /**
+     * Performs substitutions defined for this transform, and returns the result.
+     */
+    public String performSubstitutions(String inputString) {
+        String result = inputString;
+        for (Substitution substitution : getStringSubstitutions()) {
+            result = substitution.replace(result);
+        }
+        return result;
+    }
+
+    public enum NormalizeOperation {
+        SENTENCE_CASE, TITLE_CASE
+    }
+
+    /**
+     * This class holds the regex/replacement pair, and the cached compiled regex pattern.
+     */
+    public static class Substitution {
+        /** The regex string to match. */
+        public String regex;
+        /** The string that should replace regex (see effectiveReplacement below). */
+        public String replacement;
+        /** true if whitespace surrounding regex should be create or normalized to one space. */
+        public boolean normalizeSpace;
+
+        private Pattern pattern;
+        /**
+         * Replacement string that is actually used for the substitution,
+         * and set according to the value of normalizeSpace.
+         */
+        private String effectiveReplacement;
+
+        /** Empty constructor needed for persistence */
+        public Substitution() {}
+
+        public Substitution(String regex, String replacement) {
+            this(regex, replacement, false);
+        }
+
+        public Substitution(String regex, String replacement, boolean normalizeSpace) {
+            this.regex = regex;
+            this.normalizeSpace = normalizeSpace;
+            this.replacement = replacement;
+        }
+
+        /**
+         * Pre-compiles the regex pattern and determines the actual replacement string
+         * according to normalizeSpace.
+         */
+        private void initialize() {
+            if (normalizeSpace) {
+                // If normalizeSpace is set, reduce spaces before and after the regex to one space,
+                // or insert one space before and one space after if there is none.
+                // Note: if the replacement must be blank, then normalizeSpace should be set to false
+                // and whitespace management should be handled in the regex instead.
+                this.pattern = Pattern.compile(String.format("\\s*%s\\s*", regex));
+                this.effectiveReplacement = String.format(" %s ", replacement);
+            } else {
+                this.pattern = Pattern.compile(regex);
+                this.effectiveReplacement = replacement;
+            }
+        }
+
+        /**
+         * Perform the replacement of regex in the provided string, and return the result.
+         */
+        public String replace(String input) {
+            if (pattern == null) {
+                initialize();
+            }
+
+            // TODO: Study appendReplacement
```

Comment:
Can you elaborate a bit more on this TODO?

Reply:
Yes. It is related to execution speed if there are many changes to make, and you also pointed that out yourself: https://github.com/ibi-group/datatools-server/pull/371#discussion_r617124590.

Reply:
After further investigation, `Matcher.appendReplacement` is intended for processing one replacement at a time, whereas here replcements are done all at once. I removed the TODO in https://github.com/ibi-group/datatools-server/pull/371/commits/497b6bc6e33fbd6d5c5df8433d10ce8dfa126096.